File: samples/software3d/src/main/webapp/wasm-worker.js
/*
 *  Copyright 2023 Alexey Andreev.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

importScripts("wasm/software3d.wasm-runtime.js");

let instance = null;
let pendingInstanceFunctions = [];

addEventListener("message", e => {
    let data = e.data
    switch (data.type) {
        case "init":
            pendingInstanceFunctions.push(() => {
                instance.exports.initWorker(data.width, data.height, data.step, data.offset)
            });
            runPendingFunctions();
            break;
        case "frame":
            pendingInstanceFunctions.push(() => {
                instance.exports.renderFrame(data.time);
            });
            runPendingFunctions();
            break;
        case "stop":
            self.close();
            break;
    }
});

TeaVM.wasm.load("wasm/software3d.wasm", {
    installImports(o, controller) {
        o.renderer = {
            result(data, size, time) {
                let buffer = controller.instance.exports.memory.buffer.slice(data, data + size);
                self.postMessage({ data: buffer, time: time });
            }
        }
    },
}).then(teavm => {
    teavm.main([]);
    instance = teavm.instance;
    runPendingFunctions();
});


function runPendingFunctions() {
    if (instance === null) {
        return;
    }
    for (let f of pendingInstanceFunctions) {
        f();
    }
    pendingInstanceFunctions = [];
}

File: samples/software3d/src/jvmMain/kotlin/org/teavm/samples/software3d/teavm/WasmWorker.kt
/*
 *  Copyright 2023 Alexey Andreev.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package org.teavm.samples.software3d.teavm

import org.teavm.interop.Address
import org.teavm.interop.Export
import org.teavm.interop.Import
import org.teavm.samples.software3d.geometry.Matrix
import org.teavm.samples.software3d.rasterization.Raster
import org.teavm.samples.software3d.rendering.Renderer
import org.teavm.samples.software3d.scenes.geometry

fun main() {
    println("Worker started")
}

@Export(name = "initWorker")
fun init(width: Int, height: Int, step: Int, offset: Int) {
    val (scene, updaterF) = geometry()
    raster = Raster(width, Raster.calculateHeight(height, step, offset))
    updater = updaterF
    renderer = Renderer(scene, raster, offset, step).apply {
        projection = Matrix.projection(-1f, 1f, -1f, 1f, 2f, 10f)
        viewport = Matrix.translation(width / 2f, height / 2f, 0f) *
                Matrix.scale(width / 2f, width / 2f, 1f)
    }
    println("Worker initialized")
}

@Export(name = "renderFrame")
fun renderFrame(time: Double) {
    val perfStart = System.nanoTime()
    updater(time)
    renderer.render()
    val perfEnd = System.nanoTime()
    val buffer = raster.color
    sendRenderResult(Address.ofData(buffer), buffer.size * 4, (perfEnd - perfStart).toInt())
}

@Import(module = "renderer", name = "result")
external fun sendRenderResult(data: Address, dataSize: Int, time: Int)

private lateinit var raster: Raster
private lateinit var renderer: Renderer
private lateinit var updater: (Double) -> Unit
